File: bindings/java/src/test/java/org/wickra/radar/GoldenTest.java
package org.wickra.radar;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;

// Cross-language golden parity: build the radar from each committed
// golden/specs/*.json, run scan over the shared golden/events.json and read back
// the report, then assert it equals golden/expected/<spec>.json byte-for-byte. The
// binding returns the core's compact command_json string verbatim, so byte
// equality is the exact cross-language parity check. The fixtures arrive in a
// later phase; until then the test is skipped.
class GoldenTest {
    private static Path findGolden() {
        Path dir = Path.of("").toAbsolutePath();
        for (int i = 0; i < 8 && dir != null; i++) {
            Path g = dir.resolve("golden");
            if (Files.isDirectory(g.resolve("specs"))) {
                return g;
            }
            dir = dir.getParent();
        }
        return null;
    }

    @Test
    void goldenReportsAreByteIdentical() throws IOException {
        Path golden = findGolden();
        assumeTrue(golden != null, "golden fixtures not present yet");

        String events = Files.readString(golden.resolve("events.json")).strip();
        String scan = "{\"cmd\":\"scan\",\"events\":" + events + "}";
        try (Stream<Path> specs = Files.list(golden.resolve("specs"))) {
            for (Path specPath : specs.filter(p -> p.toString().endsWith(".json")).toList()) {
                String spec = Files.readString(specPath);
                String name = specPath.getFileName().toString();
                String expected = Files.readString(golden.resolve("expected").resolve(name)).strip();
                try (Radar radar = new Radar(spec)) {
                    String raw = radar.command(scan);
                    assertEquals(expected, raw.strip(), name);
                }
            }
        }
    }
}

File: bindings/node/__tests__/golden.test.js
"use strict";

// Cross-language golden parity: build the radar from each committed
// `golden/specs/*.json`, run `scan` over the shared `golden/events.json`, and
// assert the report equals `golden/expected/<spec>.json` byte-for-byte. Because
// every binding returns the core's compact `command_json` string verbatim, byte
// equality is the exact cross-language parity check. The fixtures arrive in a
// later phase; until then this test skips cleanly.

const { test } = require("node:test");
const assert = require("node:assert");
const fs = require("node:fs");
const path = require("node:path");
const { Radar } = require("../index.js");

function findGolden() {
  let dir = __dirname;
  for (let i = 0; i < 8; i++) {
    const g = path.join(dir, "golden");
    if (fs.existsSync(path.join(g, "specs"))) {
      return g;
    }
    dir = path.dirname(dir);
  }
  return null;
}

test("golden reports are byte-identical", (t) => {
  const golden = findGolden();
  if (!golden) {
    t.skip("golden fixtures not present yet");
    return;
  }
  const events = fs.readFileSync(path.join(golden, "events.json"), "utf8");
  const specDir = path.join(golden, "specs");
  for (const file of fs.readdirSync(specDir).filter((f) => f.endsWith(".json"))) {
    const spec = fs.readFileSync(path.join(specDir, file), "utf8");
    const expected = fs
      .readFileSync(path.join(golden, "expected", file), "utf8")
      .trim();
    const radar = new Radar(spec);
    const response = radar.command(
      JSON.stringify({ cmd: "scan", events: JSON.parse(events) }),
    );
    assert.strictEqual(response.trim(), expected, `mismatch for ${file}`);
  }
});
